src/eos-updater-avahi.c

```diff
@@ -0,0 +1,678 @@
+
+      queue_error_callback (discoverer,
+                            "Failed to resolve service %s: %s",
+                            name,
+                            avahi_strerror (failure));
+      return;
+    }
+
+  g_hash_table_add (discoverer->discovered_services, g_strdup (name));
+}
+
+static void
+browse_remove (EosAvahiDiscoverer *discoverer,
+               const char *name)
+{
+  guint iter;
+
+  if (g_hash_table_remove (discoverer->discovered_services, name))
+    {
+      maybe_queue_success_callback (discoverer);
+      return;
+    }
+  for (iter = 0; iter < discoverer->found_services->len; ++iter)
+    {
+      EosAvahiService *service = g_ptr_array_index (discoverer->found_services, iter);
+
+      if (g_strcmp0(service->name, name) == 0)
+        {
+          g_ptr_array_remove_index_fast (discoverer->found_services, iter);
+          break;
+        }
+    }
+}
+
+static void
+browse_discovery_finished (EosAvahiDiscoverer *discoverer)
+{
+  if (discoverer->state == EOS_AVAHI_RESOLVING_ONLY)
+    return;
+
+  discoverer->state = EOS_AVAHI_RESOLVING_ONLY;
+  maybe_queue_success_callback (discoverer);
+}
+
+static void
+browse_cb (AvahiServiceBrowser *browser,
+           AvahiIfIndex interface,
+           AvahiProtocol protocol,
+           AvahiBrowserEvent event,
+           const char *name,
+           const char *type,
+           const char *domain,
+           AvahiLookupResultFlags flags,
+           void* discoverer_ptr)
+{
+  EosAvahiDiscoverer *discoverer = discoverer_ptr;
+
+  if (discoverer->state == EOS_AVAHI_FINISHED)
+    return;
+
+  switch (event)
+    {
+    case AVAHI_BROWSER_NEW:
+      browse_new (discoverer, interface, protocol, name, type, domain);
+      break;
+
+    case AVAHI_BROWSER_REMOVE:
+      browse_remove (discoverer, name);
+      break;
+
+    case AVAHI_BROWSER_CACHE_EXHAUSTED:
+    case AVAHI_BROWSER_ALL_FOR_NOW:
+      browse_discovery_finished (discoverer);
+      break;
+
+    case AVAHI_BROWSER_FAILURE:
+      queue_error_callback (discoverer,
+                            "Avahi browser error: %s",
+                            avahi_strerror (avahi_client_errno (discoverer->client)));
+      break;
+    }
+}
+
+EosAvahiDiscoverer *
+eos_avahi_discoverer_new (GMainContext *context,
+                          EosAvahiDiscovererCallback callback,
+                          gpointer user_data,
+                          GDestroyNotify notify,
+                          GError **error)
+{
+  g_autoptr(EosAvahiDiscoverer) discoverer = NULL;
+  int failure = 0;
+
+  g_return_val_if_fail (callback != NULL, NULL);
+
+  avahi_set_allocator (avahi_glib_allocator ());
+
+  discoverer = g_object_new (EOS_TYPE_AVAHI_DISCOVERER, NULL);
+  discoverer->callback = callback;
+  discoverer->user_data = user_data;
+  discoverer->notify = notify;
+  if (context == NULL)
+    discoverer->context = g_main_context_ref_thread_default ();
+  else
+    discoverer->context = g_main_context_ref (context);
```

Comment:
Maybe, I don't know. Currently we pass a context that is actually a thread default, so the `if` is not executed, only `else`.

I think that I ended up using plain avahi instead of gobject bindings because of the main context thing - gobject bindings do not allow using custom context, they always use the global one (not a thread default one). I even filed a PR for that, but nothing happened there (https://github.com/lathiat/avahi/pull/54)